basic_nodes/src/main/java/gay/oss/gatos/basicnodes/VariableExtractionNodeType.java

```diff
@@ -28,23 +40,93 @@ public Set<NodeConnector.Input<?>> inputs(UUID nodeId, Map<String, DataBox<?>> s
+            : DataType.DATATYPE;
         return Set.of(
-            new NodeConnector.Output<>(nodeId, "output", DataType.JSONELEMENT)
+            new NodeConnector.Output<>(nodeId, "output", returnDataType.optionalOf())
         );
     }
 
     @Override
     public Map<String, CompletableFuture<DataBox<?>>> compute(Map<String, DataBox<?>> inputs, Map<String, DataBox<?>> settings) {
-        var jsonInput = DataBox.get(inputs, "input", DataType.JSONOBJECT).orElse(null);
-        var keyStr = DataBox.get(inputs, "key", DataType.STRING).orElse(null);
-        if (jsonInput == null || keyStr == null || jsonInput.get(keyStr) == null) {
-            return Map.of("output",
-                CompletableFuture.completedFuture(DataType.JSONELEMENT.create(JsonNull.INSTANCE))
+        var jsonInput = DataBox.get(inputs, "input", DataType.JSONOBJECT).orElse(new JsonObject());
+        var keyStr = DataBox.get(inputs, "key", DataType.STRING).orElse("");
+        var returnType = !settings.isEmpty() ? (ReturnType) settings.get("output_type").value() : ReturnType.STRING;
+        var value = jsonInput.get(keyStr);
+        if (value == null) {
+            return Map.of("output", CompletableFuture.completedFuture(
+                returnType.getDataType().optionalOf().create(Optional.empty()))
             );
         }
-        var value = jsonInput.get(keyStr);
-        return Map.of("output",
-            CompletableFuture.completedFuture(DataType.JSONELEMENT.create(value))
+        return switch (returnType) {
+            case INTEGER -> Map.of("output", CompletableFuture.completedFuture(
+                DataType.INTEGER.optionalOf().create(Optional.of(value.getAsInt()))));
+            case BOOLEAN -> Map.of("output", CompletableFuture.completedFuture(
+                DataType.BOOLEAN.optionalOf().create(Optional.of(value.getAsBoolean()))));
+            case STRING -> Map.of("output", CompletableFuture.completedFuture(
+                DataType.STRING.optionalOf().create(Optional.of(value.getAsString()))));
+            case JSONOBJECT -> Map.of("output", CompletableFuture.completedFuture(
+                DataType.JSONOBJECT.optionalOf().create(Optional.of(value.getAsJsonObject()))));
+            case INTEGER_LIST,
+                STRING_LIST,
+                BOOLEAN_LIST,
+                JSONOBJECT_LIST -> Map.of("output", this.handleListReturn(returnType.getDataType(), value));
+        };
+    }
+
+    private CompletableFuture handleListReturn(DataType dataType, JsonElement jsonElement) {
+        List<Object> outputList = new ArrayList<>();
+        if (jsonElement instanceof JsonArray arr && arr.size() > 0) {
+            arr.forEach(json -> outputList.add(this.jsonToReturnableType(json)));
+        } else {
+            outputList.add(this.jsonToReturnableType(jsonElement));
+        }
+        return CompletableFuture.completedFuture(dataType.optionalOf().create(
```

Comment:
No need to return optionals for list types - just return empty list if the variable doesn't exist. We don't want users to have to sometimes unwrap things _twice_.

Reply:
see updated